### PR TITLE
Fix dependency caching in CI

### LINKED ISSUE #867: Mix deps cache always misses on CI
The key which is based on mix.lock fails to point to the correct file (probably due to the working_directory change).

### PR DESCRIPTION
**Pull request by Benjamin-Philip**
This commit fixes dependency caching by doing the following:

- Upgrading from cache@v1 to cache@v3
- Using a more verbose cache path
- Changing the key format to be a bit more clearer

Caching attempts were failing as result of an invalid path, which did not account for the change in working directory.
By using a more verbose path, this issue has been eliminated.

Closes #867

### CODE REVIEW

**Comment by Benjamin-Philip** [comment]
I hate debugging CIs since this. Inability to test something on my own machine, and instead pushing to github and waiting for a runner is infuriating. I wish we had an alternative to GH Actions based on git hooks and bash scripts instead!

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -28,11 +28,13 @@ jobs:
...
+        uses: actions/cache@v3
         id: mix-cache # id to use in retrieve action
         with:
-          path: deps
-          key: v1-${{ matrix.os }}-${{ matrix.otp }}-${{ matrix.elixir }}-mix-${{ hashFiles(format('{0}{1}', github.workspace, '/mix.lock')) }}
+          path: ${{ github.workspace }}/${{ matrix.working_directory }}/deps
+          key: ${{ runner.os }}-Elixir-v${{ matrix.elixir }}-OTP-${{ matrix.otp }}-${{ hashFiles(format('{0}/{1}/mix.lock', github.workspace, matrix.working_directory)) }}
       - name: Retrieve libtorch cache
         if: ${{ matrix.working_directory == 'torchx' }}
         uses: actions/cache@v1
```
Comment (polvalente) [inline .github/workflows/ci.yml:40]: There's at least this other usage of cache that needs to be bumped. I've updated the allowed actions list to allo cache v3

Comment (Benjamin-Philip) [reply .github/workflows/ci.yml:40]: Ah yes. I had forgotten about the libtorch caches. I will update them soon.

Comment (Benjamin-Philip) [reply .github/workflows/ci.yml]: Fixed.

**Comment by Benjamin-Philip** [comment]
You will have to rerun the job for the deps to cache.

**Review by polvalente — APPROVED** [review]

**Comment by polvalente** [comment]
Thanks! It seems that both libtorch and mix deps cache are working now!

**Pull request merged**